clients/src/main/java/org/apache/kafka/common/network/PlaintextChannelBuilder.java

```diff
@@ -54,12 +54,18 @@ public void configure(Map<String, ?> configs) throws KafkaException {
     @Override
     public KafkaChannel buildChannel(String id, SelectionKey key, int maxReceiveSize,
                                      MemoryPool memoryPool, ChannelMetadataRegistry metadataRegistry) throws KafkaException {
+        PlaintextTransportLayer transportLayer = null;
         try {
-            PlaintextTransportLayer transportLayer = buildTransportLayer(key);
-            Supplier<Authenticator> authenticatorCreator = () -> new PlaintextAuthenticator(configs, transportLayer, listenerName);
+            transportLayer = buildTransportLayer(key);
+            final PlaintextTransportLayer finalTransportLayer = transportLayer;
+            Supplier<Authenticator> authenticatorCreator = () -> new PlaintextAuthenticator(configs, finalTransportLayer, listenerName);
             return buildChannel(id, transportLayer, authenticatorCreator, maxReceiveSize,
                     memoryPool != null ? memoryPool : MemoryPool.NONE, metadataRegistry);
         } catch (Exception e) {
+            // Ideally these resources are closed by the KafkaChannel but this builder should close the resources instead
+            // if an error occurs due to which KafkaChannel is not created.
+            Utils.closeQuietly(transportLayer, "transport layer for channel Id: " + id);
+            Utils.closeQuietly(metadataRegistry, "metadataRegistry");
```

Comment:
I looked into this a bit more and it's a bit odd how we're doing clean-up. `buildAndAttach` currently closes the socket channel and cancels the key. It also creates the selector channel metadata registry. Shouldn't that take care of closing the metadata registry since it actually creates it?

With regards to the transport layer, I wonder if we could move the cleanup logic to `KafkaChannel` instead. We could change `KafkaChannel` to take a `TransportLayer` supplier instead (or add a static factory method) and then have it handle the cleanup (instead of doing it in every implementation of channel builder.

Reply:
> Shouldn't that take care of closing the metadata registry since it actually creates it?

I agree, I missed it earlier. I have moved the responsibility of closing in case of errors to `Selector.buildAndAttachKafkaChannel`. I have also added a unit test that fails before this fix and passes afterwards.

> With regards to the transport layer, I wonder if we could move the cleanup logic to `KafkaChannel` instead

I am afraid that might not be possible because we are handling scenarios where we encounter an error while creating the `KafkaChannel` itself (e.g. a scenario where authenticatorCreator throws an error in the constructor). We can close it uniformly at `ChannelBuilder` but that would require a bit of refactoring involving conversion of `ChannelBuiler` interface into a class. Please let me know if you have any ideas.